SignalServiceKit/src/Devices/OWSDevice.m

```diff
@@ -5,26 +5,90 @@
 #import "OWSDevice.h"
 #import "NSDate+OWS.h"
 #import "OWSError.h"
+#import "TSStorageManager.h"
 #import "YapDatabaseConnection.h"
 #import "YapDatabaseTransaction.h"
 #import <Mantle/MTLValueTransformer.h>
 
 NS_ASSUME_NONNULL_BEGIN
 
-static MTLValueTransformer *_millisecondTimestampToDateTransformer;
 uint32_t const OWSDevicePrimaryDeviceId = 1;
+NSString *const kTSStorageManager_OWSDeviceCollection = @"kTSStorageManager_OWSDeviceCollection";
+NSString *const kTSStorageManager_MayHaveLinkedDevices = @"kTSStorageManager_MayHaveLinkedDevices";
+
+@interface OWSDeviceManager ()
+
+@property (atomic, nullable) NSNumber *mayHaveLinkedDevicesCached;
+
+@end
+
+#pragma mark -
+
+@implementation OWSDeviceManager
+
++ (instancetype)sharedManager
+{
+    static OWSDeviceManager *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        instance = [[self alloc] initDefault];
+    });
+    return instance;
+}
+
+- (instancetype)initDefault
+{
+    return [super init];
+}
+
+- (BOOL)mayHaveLinkedDevices:(YapDatabaseConnection *)dbConnection
+{
+    OWSAssert(dbConnection);
+
+    @synchronized(self)
+    {
+        if (!self.mayHaveLinkedDevicesCached) {
+            self.mayHaveLinkedDevicesCached = @([dbConnection boolForKey:kTSStorageManager_MayHaveLinkedDevices
+                                                            inCollection:kTSStorageManager_OWSDeviceCollection
+                                                            defaultValue:YES]);
+        }
+
+        return [self.mayHaveLinkedDevicesCached boolValue];
+    }
+}
+
+- (void)setMayHaveLinkedDevices:(BOOL)value dbConnection:(YapDatabaseConnection *)dbConnection
+{
+    OWSAssert(dbConnection);
+
+    @synchronized(self)
+    {
+        self.mayHaveLinkedDevicesCached = @(YES);
```

Comment:
shouldn't this be ` self.mayHaveLinkedDevicesCached = @(value);`?

Reply:
Other than that, LGTM! Thanks for considering my voluminous feedback.

Reply:
Done. Thanks for repeatedly saving me from myself.